reportssql/testing_report_util_functions.sql

```diff
@@ -458,6 +458,201 @@ BEGIN
+        AND o.value_coded = (SELECT concept.concept_id FROM concept WHERE concept.uuid = positiveUuid)
+    ORDER BY o.date_created DESC
+    LIMIT 1;
+
+    RETURN (result);
+END$$
+DELIMITER ;
+
+-- getDateOfVirologicTest
+
+DROP FUNCTION IF EXISTS getDateOfVirologicTest;
+
+DELIMITER $$
+CREATE FUNCTION getDateOfVirologicTest(
+    p_patientId INT(11),
+    p_startDate DATE,
+    p_endDate DATE) RETURNS TINYINT(1)
+    DETERMINISTIC
+BEGIN
+
+    DECLARE dateAtVirologicHIVTestLabForm DATE;
+    DECLARE dateAtVirologicHIVTestElis DATE;
+    DECLARE dateAtVirologicHIVTest DATE;
+
+    SET dateAtVirologicHIVTestLabForm = getDateOfVirologicHIVTestFromLabForm(p_patientId, p_startDate, p_endDate);
+    SET dateAtVirologicHIVTestElis = getDateOfVirologicHIVTestFromElis(p_patientId, p_startDate, p_endDate);
+
+    IF dateAtVirologicHIVTestLabForm IS NULL AND dateAtVirologicHIVTestElis IS NULL THEN
+        RETURN FALSE;
+    ELSEIF dateAtVirologicHIVTestElis IS NOT NULL THEN
+        SET dateAtVirologicHIVTest = dateAtVirologicHIVTestElis;
+    ELSE
+        SET dateAtVirologicHIVTest = dateAtVirologicHIVTestLabForm;
+    END IF;
+
+    RETURN (dateAtVirologicHIVTest);
+END$$
+DELIMITER ;
+
+-- patientWasNotEnrolledToHIVProgramBeforeVirologicTest
+
+DROP FUNCTION IF EXISTS patientWasNotEnrolledToHIVProgramBeforeVirologicTest;
+
+DELIMITER $$
+CREATE FUNCTION patientWasNotEnrolledToHIVProgramBeforeVirologicTest(
+    p_patientId INT(11),
+    p_startDate DATE,
+    p_endDate DATE) RETURNS TINYINT(1)
+    DETERMINISTIC
+BEGIN
+
+    DECLARE dateOfVirologicHIVTest DATE;
+
+    SET dateOfVirologicHIVTest = getDateOfVirologicTest(p_patientId, p_startDate, p_endDate);
+
+    RETURN (NOT patientHasEnrolledIntoHivProgramBefore(p_patientId, dateOfVirologicHIVTest));
+END$$
+DELIMITER ;
+
+-- patientWasNotInitiatedToARVBeforeVirologicTest
+
+DROP FUNCTION IF EXISTS patientWasNotInitiatedToARVBeforeVirologicTest;
+
+DELIMITER $$
+CREATE FUNCTION patientWasNotInitiatedToARVBeforeVirologicTest(
+    p_patientId INT(11),
+    p_startDate DATE,
+    p_endDate DATE) RETURNS TINYINT(1)
+    DETERMINISTIC
+BEGIN
+
+    DECLARE dateOfVirologicHIVTest DATE;
+
+    SET dateOfVirologicHIVTest = getDateOfVirologicTest(p_patientId, p_startDate, p_endDate);
+
+    RETURN (NOT patientHasStartedARVTreatmentBefore(p_patientId, dateOfVirologicHIVTest));
+END$$
+DELIMITER ;
+
+-- patientNotTakingARVAtDateOfVirologicTest
+
+DROP FUNCTION IF EXISTS patientNotTakingARVAtDateOfVirologicTest;
+
+DELIMITER $$
+CREATE FUNCTION patientNotTakingARVAtDateOfVirologicTest(
+    p_patientId INT(11),
+    p_startDate DATE,
+    p_endDate DATE) RETURNS TINYINT(1)
+    DETERMINISTIC
+BEGIN
+
+    DECLARE dateOfVirologicHIVTest DATE;
+
+    SET dateOfVirologicHIVTest = getDateOfVirologicTest(p_patientId, p_startDate, p_endDate);
+
+    RETURN (NOT patientWasOnARVTreatmentOrHasPickedUpADrugWithinReportingPeriod(p_patientId, dateOfVirologicHIVTest, dateOfVirologicHIVTest, 0));
```

Comment:
@tmvumbi2 This period doesn't seem right, busy checking with Richard what should we use.